File: src/CircleAI.Speech.Cloud/GoogleSpeechSynthesizer.cs
// GoogleSpeechSynthesizer.cs
//
// (3.3.0) ISpeechSynthesizer backed by Google Cloud TTS v1
// /v1/text:synthesize. API-key auth; returns base64 LINEAR16 audio.

using System;
using System.Net.Http;
using System.Text;
using System.Text.Json;
using System.Threading;
using System.Threading.Tasks;
using Microsoft.Extensions.Logging;
using Microsoft.Extensions.Logging.Abstractions;

namespace CircleAI.Speech.Cloud;

/// <summary>(3.3.0) Google-backed <see cref="ISpeechSynthesizer"/>.</summary>
public sealed class GoogleSpeechSynthesizer : ISpeechSynthesizer
{
    private readonly HttpClient _http;
    private readonly GoogleTtsOptions _options;
    private readonly ILogger _logger;

    public GoogleSpeechSynthesizer(HttpClient http, GoogleTtsOptions options, ILogger<GoogleSpeechSynthesizer>? logger = null)
    {
        _http    = http    ?? throw new ArgumentNullException(nameof(http));
        _options = options ?? throw new ArgumentNullException(nameof(options));
        _logger  = (ILogger?)logger ?? NullLogger.Instance;

        if (_http.BaseAddress is null) _http.BaseAddress = options.BaseAddress;
    }

    public string BackendId    => "google-tts";
    public bool   IsConfigured => !string.IsNullOrWhiteSpace(_options.ApiKey);

    public async ValueTask<SynthesisResult> SynthesizeAsync(
        string            text,
        string?           voiceId      = null,
        string?           languageHint = null,
        CancellationToken ct           = default)
    {
        if (!IsConfigured) return Empty();

        var voice = string.IsNullOrWhiteSpace(voiceId) ? _options.DefaultVoiceName : voiceId;
        var lang  = string.IsNullOrWhiteSpace(languageHint) ? _options.LanguageCode : languageHint;

        var body = $$"""
            {
              "input": { "text": {{JsonSerializer.Serialize(text)}} },
              "voice": {
                "languageCode": "{{lang}}",
                "name": "{{voice}}"
              },
              "audioConfig": {
                "audioEncoding": "LINEAR16",
                "sampleRateHertz": {{_options.PcmSampleRateHz}}
              }
            }
            """;

        var path = $"/v1/text:synthesize?key={Uri.EscapeDataString(_options.ApiKey!)}";
        using var resp = await _http.PostAsync(path, new StringContent(body, Encoding.UTF8, "application/json"), ct).ConfigureAwait(false);
        if (!resp.IsSuccessStatusCode)
        {
            _logger.LogWarning("Google TTS returned {Status}", resp.StatusCode);
            return Empty();
        }

        using var doc = await JsonDocument.ParseAsync(
            await resp.Content.ReadAsStreamAsync(ct).ConfigureAwait(false),
            cancellationToken: ct).ConfigureAwait(false);

        if (!doc.RootElement.TryGetProperty("audioContent", out var ac)) return Empty();
        var b64   = ac.GetString();
        if (string.IsNullOrEmpty(b64)) return Empty();

        var bytes = Convert.FromBase64String(b64);
        // Google returns a WAV envelope — strip it.
        var pcm   = StripWavHeader(bytes);
        var samples = pcm.Length / 2;
        return new SynthesisResult(
            AudioPcm16Mono: pcm,
            SampleRateHz:   _options.PcmSampleRateHz,
            Duration:       TimeSpan.FromSeconds((double)samples / _options.PcmSampleRateHz));
    }

    /// <summary>Strip a 44-byte WAV header if present.</summary>
    private static byte[] StripWavHeader(byte[] data)
    {
        if (data.Length > 44 && data[0] == 'R' && data[1] == 'I' && data[2] == 'F' && data[3] == 'F')
        {
            var stripped = new byte[data.Length - 44];
            Array.Copy(data, 44, stripped, 0, stripped.Length);
            return stripped;
        }
        return data;
    }

    private static SynthesisResult Empty() =>
        new(ReadOnlyMemory<byte>.Empty, SampleRateHz: 0, Duration: TimeSpan.Zero);
}

File: python/src/circle_ai/speech/cloud/google_speech_synthesizer.py
# speech/cloud/google_speech_synthesizer.py
#
# Port of CircleAI.Speech.Cloud/GoogleSpeechSynthesizer.cs (C# — the EXACT spec).
#
# (3.3.0) ISpeechSynthesizer backed by Google Cloud TTS v1 /v1/text:synthesize.
# API-key auth; input.text + voice + audioConfig LINEAR16 JSON body. The response
# carries base64 "audioContent" — a WAV envelope that the C# strips back to raw
# PCM. Fail-soft on missing key / non-2xx.
#
# The C# drives HttpClient directly; the Python port injects the shared
# circle_ai.integration.http.IHttpFetcher. The JSON body rides body_json (the
# fetcher serialises it — note the C# uses JsonSerializer.Serialize(text) for the
# input string, which is exactly json.dumps of a str, so body_json is faithful);
# the response is read off resp.json() then base64-decoded + WAV-stripped.

from __future__ import annotations

import base64
import logging
from datetime import timedelta
from typing import Optional

from ...integration.http import HttpRequest, IHttpFetcher
from ..contracts import ISpeechSynthesizer, SynthesisResult
from ._audio_http import combine_uri, is_null_or_whitespace, strip_wav_header
from .options import GoogleTtsOptions

_logger = logging.getLogger("CircleAI.Speech.Cloud.GoogleSpeechSynthesizer")


def _empty() -> SynthesisResult:
    return SynthesisResult(b"", 0, timedelta(0))


def _escape(value: str) -> str:
    from urllib.parse import quote

    return quote(value, safe="-_.~")


class GoogleSpeechSynthesizer(ISpeechSynthesizer):
    """(3.3.0) Google-backed :class:`ISpeechSynthesizer`.

    Mirrors ``CircleAI.Speech.Cloud.GoogleSpeechSynthesizer``.
    """

    def __init__(
        self,
        http: IHttpFetcher,
        options: GoogleTtsOptions,
        logger: Optional[logging.Logger] = None,
    ) -> None:
        if http is None:
            raise ValueError("http must not be None")
        if options is None:
            raise ValueError("options must not be None")
        self._http = http
        self._options = options
        self._logger = logger if logger is not None else _logger

    @property
    def backend_id(self) -> str:
        return "google-tts"

    @property
    def is_configured(self) -> bool:
        return not is_null_or_whitespace(self._options.api_key)

    async def synthesize_async(
        self,
        text: str,
        voice_id: Optional[str] = None,
        language_hint: Optional[str] = None,
        ct: object = None,
    ) -> SynthesisResult:
        if not self.is_configured:
            return _empty()

        voice = self._options.default_voice_name if is_null_or_whitespace(voice_id) else voice_id
        lang = self._options.language_code if is_null_or_whitespace(language_hint) else language_hint

        body = {
            "input": {"text": text},
            "voice": {"languageCode": lang, "name": voice},
            "audioConfig": {
                "audioEncoding": "LINEAR16",
                "sampleRateHertz": self._options.pcm_sample_rate_hz,
            },
        }

        path = f"/v1/text:synthesize?key={_escape(self._options.api_key or '')}"
        resp = await self._http.send_async(
            HttpRequest(
                method="POST",
                url=combine_uri(self._options.base_address, path),
                body_json=body,
            )
        )
        if not resp.is_success:
            self._logger.warning("Google TTS returned %s", resp.status_code)
            return _empty()

        doc = resp.json()
        if not isinstance(doc, dict) or "audioContent" not in doc:
            return _empty()
        b64 = doc.get("audioContent")
        if not b64 or not isinstance(b64, str):
            return _empty()

        try:
            raw = base64.b64decode(b64)
        except (ValueError, TypeError):
            return _empty()
        # Google returns a WAV envelope — strip it.
        pcm = strip_wav_header(raw)
        samples = len(pcm) // 2
        return SynthesisResult(
            pcm,
            self._options.pcm_sample_rate_hz,
            timedelta(seconds=samples / self._options.pcm_sample_rate_hz),
        )


__all__ = ["GoogleSpeechSynthesizer"]
